File: fizzbuzz.cpp
#include <iostream>
#include <stdio.h>
#include <stdexcept>
#include "fizzbuzz.h"

using namespace std;

int main()
{
    int iNum = 0;

    try
    {
        iNum = get_num();
        do_fb(iNum);
    }
    catch(const invalid_argument& e)
    {
        printf("ERROR: %s", e.what());
    }
    
    return 0;
}

// -------------------------------------------------------------------

/**
 * Checks if a string contains only numeric characters.
 *
 * This function iterates over each character in the input string and checks if it is a digit.
 *
 * @param szStr_ The string to check.
 * @return `true` if the string contains only numeric characters, `false` otherwise.
 *
 * @note This function considers an empty string to be numeric.
 */
bool check_number(string szStr_)
{
    for (int i = 0; i < szStr_.length(); i++)
    {
        if (!isdigit(szStr_[i]))
            return false;
    }
   return true;
}

// -------------------------------------------------------------------

/**
 * Prompts the user to enter a numeric string, then converts it to an integer.
 *
 * This function prompts the user to enter a string and reads it from standard input using scanf.
 * It then converts the string to an integer using stoi, and returns the result.
 *
 * @throws std::invalid_argument if the input string is not a valid integer or is not positive.
 * @return The integer value entered by the user.
 *
 * @note This function does not handle input errors gracefully, and may exit the program if invalid input is provided.
 */
int get_num()
{
    int iResult = 0;
    char cTemp[128];
    string szNum = "";

    printf("Enter a positive integer number: ");

    iResult = scanf("%127s", &cTemp);

    szNum = cTemp;

    if(iResult != 1 || !check_number(szNum))
        throw invalid_argument("Is not an integer or positive");
    else
        return stoi(szNum);
}

// -------------------------------------------------------------------

/**
 * Computes the FizzBuzz string for a single number.
 *
 * This function takes an input integer and computes the FizzBuzz string for that number.
 * If the number is divisible by 3, "Fizz" is appended to the output string.
 * If the number is divisible by 5, "Buzz" is appended to the output string.
 * If the number is not divisible by either 3 or 5, the input number is converted to a string and returned.
 *
 * @param iNum_ The number to compute the FizzBuzz string for.
 * @return The computed FizzBuzz string.
 */
string do_single_fb(int iNum_)
{
    string szOutput = "";

    if((iNum_ % 3) == 0)
    {
        szOutput += "Fizz";
    }
    if((iNum_ % 5) == 0)
    {
        szOutput += "Buzz";
    }
    if(szOutput == "")
    {
        szOutput = to_string(iNum_);
    }

    return szOutput;
}

// -------------------------------------------------------------------

/**
 * Computes and prints the FizzBuzz string for numbers 1 to the given input.
 *
 * This function takes an input integer and computes the FizzBuzz string for all numbers from 1 to the input.
 * For each number, the `do_single_fb` function is called to compute the corresponding FizzBuzz string,
 * which is then printed to standard output using printf.
 *
 * @param iNum_ The maximum number to compute the FizzBuzz string for.
 */
void do_fb(int iNum_)
{
    string szOutput = "";
    for(int i=1; i<iNum_+1; i++)
    {
        szOutput = do_single_fb(i).data();
        printf("%s\n", szOutput.c_str());
    }
}

File: fizzbuzz.java
import java.util.Scanner;

public class FizzBuzz {
    /**
     * Main entry point of the program.
     *
     * This function is the main entry point of the program. It prompts the user to enter a positive integer number
     * using the `get_num` function, and generates a "FizzBuzz" sequence for that number using the `do_fb` function.
     * If an exception is thrown during the execution of the program, such as a NumberFormatException, the function
     * catches the exception and prints an error message to the console.
     *
     * @throws NumberFormatException if the entered value is not a non-negative integer
     */
    public static void main(String[] args) throws NumberFormatException {

        int iNum = 0;

        try {
            iNum = get_num();
            do_fb(iNum);
        } catch (Exception e) {
            System.out.println("ERROR: Input is not an integer or positive.");
        }
    }

    // -------------------------------------------------------------------
    /**
     * Prompts the user to enter a positive integer from the console, and returns the entered value.
     *
     * This function prompts the user to enter a positive integer number from the console. The function then
     * reads the entered value as an integer, and checks whether the value is positive using the `check_number`
     * function. If the entered value is not a positive integer, the function throws a NumberFormatException.
     *
     * @throws NumberFormatException if the entered value is not a positive integer
     * @return the positive integer entered by the user
     */
    public static int get_num() {

        System.out.print("Enter a positive integer number: ");

        Scanner in = new Scanner(System.in);

        // Will throw exception if number is float
        int iNum = in.nextInt();
        in.close();

        if(!check_number(iNum))
            throw new NumberFormatException();

        return iNum;
    }

    // -------------------------------------------------------------------

    /**
     * Checks if a given integer is negative.
     *
     * This function returns true if the given integer is positive (i.e., greater than or equal to zero),
     * and false otherwise.
     *
     * @param iNum_ the integer to be checked for postitivity
     * @return true if the given integer is positive, false otherwise
     */
    public static boolean check_number(int iNum_)
    {
        if(iNum_ < 0) {
            return false;
        }

        return true;
    }

    // -------------------------------------------------------------------

    /**
     * Returns a string representation of a given integer, according to the "FizzBuzz" rules.
     *
     * The "FizzBuzz" rules state that:
     * - If the number is divisible by 3, the string "Fizz" is added to the output.
     * - If the number is divisible by 5, the string "Buzz" is added to the output.
     * - If the number is not divisible by either 3 or 5, the number itself is converted to a string and added to the output.
     *
     * @param iNum_ the integer to be converted to a string according to the "FizzBuzz" rules
     * @return a string representation of the given integer, according to the "FizzBuzz" rules
     */
    public static String do_single_fb(int iNum_) {

        String szOutput = "";

        if((iNum_ % 3) == 0)
        {
            szOutput += "Fizz";
        }
        if((iNum_ % 5) == 0)
        {
            szOutput += "Buzz";
        }
        if(szOutput == "")
        {
            szOutput = Integer.toString(iNum_);
        }

        return szOutput;
    }

    // -------------------------------------------------------------------

    /**
     * Prints a sequence of "FizzBuzz" strings to the console, for integers from 1 to a given number.
     *
     * This function calls the do_single_fb function to generate "FizzBuzz" strings for each integer
     * from 1 to the given number. The "FizzBuzz" strings are printed to the console.
     *
     * @param iNum_ the number of integers to generate "FizzBuzz" strings for
     */
    public static void do_fb(int iNum_) {

        for(int i=1; i<iNum_+1; i++) {
            System.out.println(do_single_fb(i));
        }
    } 
}
